src/Language/PureScript/Scheme/CodeGen/Transpiler.hs

```diff
@@ -107,23 +120,31 @@ caseToScheme values caseAlternatives =
     condClauses (valuesAndBinders, Right result) = 
       [(test valuesAndBinders, bindersToResult valuesAndBinders result)]
 
-    -- Emit the test for a cond clause. In scheme: (cond (test1 result1) ...)
-    -- If we're handling multiple values and binders for a single result
-    -- we have to wrap the various resulting test into an `and' expression.
+    -- Emit the test for a cond clause.
     -- In scheme: (cond ((and test1a test1b) result1) ...)
-    test [(value, binder)] = binderToTest value binder
+    -- Always wrap the test in an `and' form since binderToTest has to return a
+    -- list because when applied to a ConstructorBinder it returns a list of AST.
+    -- We could have generated an `and' expression in such case in binderToTest,
+    -- but it's easier to optimize (and foo) -> foo.
+    -- TODO: Optimize (and foo) into foo.
     test valueAndBinders =
-      Application (Identifier "and")
-                  (map (\(value, binder) -> binderToTest value binder)
-                       valueAndBinders)
+      and_ (concatMap (\(value, binder) -> binderToTest (exprToScheme value)
+                                                        binder)
+           valueAndBinders)
 
     -- Emit a cond clause test for each value and binder.
     -- E.g.: (value: 23, binder: 42) -> (= 23 42)
-    binderToTest :: Expr Ann -> Binder Ann -> AST
+    binderToTest :: AST -> Binder Ann -> [AST]
+    binderToTest _value (NullBinder _ann) = [t]
     binderToTest value (LiteralBinder _ann (NumericLiteral (Left integer))) =
-      Application (Identifier "=") [exprToScheme value, IntegerLiteral integer]
-    binderToTest _value (VarBinder _ann _ident) = Identifier "#t"
-    binderToTest _value (NullBinder _ann) = Identifier "#t"
+      [eq [value, IntegerLiteral integer]]
+    binderToTest _value (VarBinder _ann _ident) = [t]
+    binderToTest value (ConstructorBinder _ann _typeName constructorName binders) =
+      (:) (eqQ (car value)
+               (quote (Identifier (runProperName (disqualify constructorName)))))
+          (concatMap (\(i, b) -> (binderToTest (vectorRef value (IntegerLiteral i)) b))
+                     (zip [1 .. (genericLength binders)]
```

Comment:
Off by one.

Reply:
Totally, also it looks like you're using this in multiple places, so it would be safer to have a helper, something like
```haskell
mapIndexed :: (a -> Int -> b) -> [a] -> [b]
mapIndexed f l = zipWith f l [0..]
```
Note that you don't need to specify the length - Haskell has lazy lists, and the `zip` will take values until your collection ends. Magic :slightly_smiling_face: